test/testlib/minikube_utilities.go

```diff
@@ -29,21 +30,86 @@ import (
 var teardownLists = make(map[string][]func())
+var diagnosticTeardownLists = make(map[string][]func())
+
+/** Exported var - initialised from the EnvVar, but can be reset in code if desired */
+var AlwaysRunDiagnosticTeardowns = strings.EqualFold(os.Getenv(ALWAYS_RUN_DIAGNOSTIC_TEARDOWNS), "true")
 
 /**
- * add a teardown function to the named list - for later execution
+ * add a teardown function to the named list - for deferred execution.
+ *
+ * The teardown functions are called in reverse order of insertion, by a call to Teardown(name).
+ *
+ * The typical idiom is:
+ * <pre>
+ *   testlib.AddTeardown("DATABASE", func() { ...})
+ *   // possibly more testlib.AddTeardown("DATABASE", func() { ... })
+ *   defer testlib.Teardown("DATABASE")
+ * <pre>
  */
 func AddTeardown(name string, teardownFunc func()) {
 	teardownLists[name] = append(teardownLists[name], teardownFunc)
 }
 
+/**
+ * add a diagnostic teardown func to be called before any other teardowns in the named list - to aid diagnostics/debugging.
+ * This allows a diagnostic teardown to do such things as:
+ * <ul>
+ *   <li>Generate logging and debug information immediately prior to resurce teardown
+ *   <li>call time.Sleep() to allow inspection and/or debugging of the exit state before teardown.
+ * <ul>
+ *
+ * NOTE: it is generally undesirable to add multiple diagnostic teardowns that sleep - so it would usually be best to
+ * add any Sleep() debug teardown to the innermost teardown list.
+ * Nonetheless, there are use-cases where multiple Sleep() teardowns are useful - to allow inspecting different
+ * intermediate states.
+ */
+func AddDiagnosticTeardown(name string, condition interface{}, teardownFunc func()) {
+	tdfunc := func() {
+		shouldIdoIt := AlwaysRunDiagnosticTeardowns
+
+		if !shouldIdoIt {
+			switch c := condition.(type) {
+			case *testing.T:
+				shouldIdoIt = c.Failed()
+
+			case func() bool:
+				shouldIdoIt = c()
+
+			case bool:
+				shouldIdoIt = c
+
+			default:
+				shouldIdoIt = c != nil
+			}
+		}
+
+		if shouldIdoIt {
+			teardownFunc()
+		}
+	}
+
+	diagnosticTeardownLists[name] = append(diagnosticTeardownLists[name], tdfunc)
+}
+
 /**
  * Call the stored teardown functions in the named list, in the correct order (last-in-first-out)
+ *
+ * NOTE: Any DIAGNOSTIC teardowns - those added with AddDiagnosticTeardown() for this name - are called BEFORE any other teardowns for this name.
+ *
+ * The typical use of Teardown is with a deferred call:
+ * defer testlib.Teardown("SOME NAME")
+ * See: testlib.AddTeardown(); testlib.AddDiagnosticTeardown()
  */
 func Teardown(name string) {
+	// ensure both list and diagnostic list are removed.
+	defer func() { delete(diagnosticTeardownLists, name) }()
+	defer func() { delete(teardownLists, name) }()
+
 	list := teardownLists[name]
-	delete(teardownLists, name)
+	list = append(list, diagnosticTeardownLists[name]...) // append any diagnostic funcs - so they are called FIRST
```

Comment:
my suggestion was to conditionalize the diagnostic teardown here. Something like:
`func Teardown(t *testing.T, name string) {
	list := teardownLists[name]

	delete(teardownLists, name)

	if t.Failed() || AlwaysRunDiagnosticTeardowns {
		list = append(list, diagnosticTeardownLists[name]...) // append any diagnostic funcs - so they are called FIRST
	}

	delete(diagnosticTeardownLists, name)
	
	for _, lmbd := range list {
		lmbd()
	}
}`

Reply:
> somehow your branch is colliding with your previous dep->modules change
Seems I've somehow messed up the rebase of this branch onto master.

If that seems like a real problem, I'll just create (yet another) new branch,and just port these changes across manually - which is what the rebase was supposed to do automatically... :/